### ISSUE TITLE
Allow customization of entryPoint in DockerJavaApplicationPlugin

### ISSUE THREAD

**Issue by mwhipple** [bmuschko/gradle-docker-plugin#358]
I'd like to add an `entryPoint` to the application extension which would allow pre-empting the current `determineEntryPoint` logic. I should be able to get a PR together (likely tomorrow) unless there are any objections.

**Comment by orzeh** [comment]
@mwhipple any PRs are welcome. please take a look at #306 which is about the same issue.

**Comment by mwhipple** [comment]
OK, cool. I looked at #306 but it initially seemed different enough to keep separate but now I'm thinking both could be addressed cleanly if the extension point is a more generalized closure delegated to the Dockerfile task
```
launchCommand {
  entryPoint('run.sh')
}
```
(naming TBD)

**Comment by orzeh** [comment]
@mwhipple sounds good!

**Comment by bmuschko** [comment]
@mwhipple I was just having a look at this code. Could you explain the use case for this change (with a concrete, real-world example)? Why do you require customization of the default entrypoint the plugin provides?

**Comment by mwhipple** [comment]
Sure (this plugin seems to be getting a fairly thorough auditing :)). My use of this feature involves not so much customizing the default entry point as augmenting the startup in a minimally invasive way by having a script that performs some setup work and then exec's the original entrypoint script. These are cases where the application itself adheres to the conventions of this plugin, but some launch time preparation is required due to the deployment environment. Some concrete examples:

 - Augmenting the environment (variables) with the host address to allow communication with other locally running containers (such as the Datadog agent)
- Performing decryption of sensitive data using a tool such as [shush](https://github.com/realestate-com-au/shush)

Ultimately things that should likely be the responsibility of the system that is launching the containers but until that ability (such as initcontainers) is refined and widespread a script which enriches the container launch provides a contained solution.

**Comment by mwhipple** [comment]
I had plans to add some groovy doc the composite instruction since it acquired a link from other docs. Let me know if that is desired.

**Comment by bmuschko** [comment]
@mwhipple Thanks for your explanation. Let me reiterate what I understand. You'd like to keep the start script in place produced by the application plugin. You'd want to add another shell script with additional setup that in turn calls the original start script. Is it correct to assume that you also add a `ADD` or `COPY` instruction to the Dockerfile. If yes, how do you do it? The [user guide example](http://bmuschko.github.io/gradle-docker-plugin/#adding_custom_instructions_to_dockerfile) and the [existing functional test](https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/functTest/groovy/com/bmuschko/gradle/docker/DockerJavaApplicationPluginFunctionalTest.groovy#L112) only show a `run.sh` file but doesn't explain how you tell the container where it is coming from.

Instead of making the API of the Java application plugin more complex I am wondering if you considered one of these options:

1. Making these customizations in the application plugin. From a "responsibility" perspective I'd rather see it there than making the API of the Docker plugin more complex. See the API of [CreateStartScripts](https://docs.gradle.org/current/dsl/org.gradle.jvm.application.tasks.CreateStartScripts.html) for more information. This would replace the existing start script with one that you provide.

```
startScripts {
    unixStartScriptGenerator.template = resources.text.fromFile('customUnixStartScript.txt')
}
```

2. Modifying the instructions in `dockerDistTar`, the task that generates the Dockerfile instructions.

```
int index = dockerDistTar.instructions.findIndexOf { it.keyword == 'ENTRYPOINT' }
dockerDistTar.instructions[index] = new EntryPointInstruction('run.sh')
```

Two reasons why I am asking these questions:

1. I didn't fully understand the use case. I believe the user guide doesn't do a good job at the moment in explaining the reasoning. As a user I didn't quite understand the need for it after reading the paragraph.
2. The existing `exec` functionality makes the code harder to maintain, the DSL more complex to understand and requires additional testing. It also only handles the case of `CMD` and `ENTRYPOINT`. What if you wanted to make other customizations? Then you'd have to go back to `dockerDistTar` or make the API of `CompositeExecInstruction` even more complex.

Bottom line: I'd love to remove this functionality if there's a easier solution to the problem. Let me know if I am missing something.

**Comment by mwhipple** [comment]
The way that the custom script was added to the project was by adding it into the dist dir before packaging with something like:
```
task('copyEntrypointScript', type: Copy) {
    from "$rootDir/src/main/scripts/entrypoint.sh"
    into { "$buildDir/scripts" }
}
dockerBuildImage.dependsOn 'copyEntrypointScript'
```
(I try to avoid scattering packaging info between the Dockerfile and the build system)

For the further questions I did consider the start script but that seemed to be unnecessarily cumbersome since my concern wasn't around the functionality actually provided by that script: so the idea of leaving the existing logic in place and adding to it seemed to avoid having to touch additional pieces and incur the promise of future maintenance costs.

For the second question I think that's a judgement call for you guys :). I use customization of the `dockerDistTar` task in some projects, and this functionality could certainly be replaced by a recipe to that effect if it introduces undue maintenance overhead. At the time it seemed like it would be a common use case (based on virtually no data other than the other open issue) so extending the DSL of that plugin seemed like it would align with the goal of the convenience provided. To me this largely breaks down into what use cases are handled by the application plugin and which should fall out to manually stitching together the RemoteApiPlugin, where the solution of modifying the dockerDistTar task falls into the latter while allowing the app plugin to provide the major pieces (i.e. it still requires a more in-depth knowledge of the plugin).

So in short the removal will IMO significantly raise the convenience factor of this plugin for anyone needing similar functionality but I have no idea how many people that is and if the maintenance cost is too high then a recipe would certainly make more sense.

**Comment by mwhipple** [comment]
Regaring the limitations of exec: it seems like another variation of the same question: where is the line between the opinionated DSL and the more generalized framework? (there's some gradle design principle like that which I'm sure you know  ;) but I don't remember the specific wording). The framework is there, the opinions are yours to have.

**Comment by bmuschko** [comment]
> So in short the removal will IMO significantly raise the convenience factor of this plugin for anyone needing similar functionality but I have no idea how many people that is and if the maintenance cost is too high then a recipe would certainly make more sense.

I'll think about this more over the weekend. Overall I feel that we are trying to solve a shortcoming of the Gradle core application plugin. You'd have the same problem if you wanted to customize the distribution produced by the application plugin independent on whether it should run in a container or not. You basically want to say "before running the program set up the environment". Creating another shell script that call the original script is somewhat a "workaround" for the fact that the original script is not customizable easily.

> where is the line between the opinionated DSL and the more generalized framework?

I don't think I quite understand the question. The Java application plugin is extremely opinionated. It introduces its own DSL through the extension `DockerJavaApplication` and provides its own set of tasks and instructions. What exactly do you mean by "generalized framework"?

**Comment by mwhipple** [comment]
by "generalized framework"  I just meant that all the essential pieces are provided to be wired together imperatively with the appropriate logic (framework was probably a bad word choice and I may have been confusing or abusing the underlying principle), and that bit was referring to the decisions around where to add conventions and more declarative approaches, but I don't think this track of conversation is worth going further down.

I fully agree with the first paragraph, I may have missed earlier allusions to that idea. Adding some form of placeholder hooks in the startup scripts seems like it would be simple and provide equivalent extensibility in a way that is closer to the rest of the artifact packaging while avoiding any concerns I could think of.

**Comment by mwhipple** [comment]
I'll add a bit more input if the short term solution is still a question. I think we're saying the same thing from different angles regarding capabilities vs. contentions: as you just said the capabilities are there so whether it makes sense to have more declarative support on top of the conventions or just potentially provide guidance on a more manual approach is somewhat discretionary (that's the line I was referring to, and IMO that's the same discussion around the current limitations of exec).

Based on my very limited perspective and speculation, here's a couple extra 2 cents in case it provides any additional accurate context.

I'd guess you have some ways to try to avoid the curse of knowledge regarding using gradle, but beyond the "convenience" factor there's also the comfort level for most developers that are primarily gradle consumers. Using the provided DSL is straightforward but in terms of going outside of that box and working on the underlying objects the comfort level drops of dramatically and normally so does the cleanliness of the solution implemented.

On the other hand, both of the use cases I mentioned are the types of changes that would be deployment environment or organization specific rather than being application specific. Therefore these would be a natural fit for an organization local plugin which implements the organization specific conventions (such as the entrypoint script itself), and that addresses the previous point and wraps up this type of customization in a way that how it is done is fairly moot. So this would be unlikely to be an issue in any shops with a decent amount of gradle expertise or with dedicated toolsmiths (or where the issues are addressed elsewhere such as the deployment environment itself).

**Comment by mwhipple** [comment]
This approach was likely also from a preference to avoid after the fact mutation (which could certainly cause tension with convention). I can play around with that route tomorrow (replacing the instruction outside of the plugin) since the resulting recipe would likely be really simple.

**Comment by bmuschko** [comment]
@mwhipple I am mostly concerned about the things below. The biggest concern for me is 1. and 2.

1. Based on the existing documentation it wasn't clear to me why you'd even want to use `exec`. It doesn't describe the use case and the _why_. We'll definitely need to improve on that. I am sure that other users will have the same issue.
2. The functional test and the code snippet in the user guide don't explain what `run.sh` is and how it is being added to the image. This is also a missing piece of documentation. Users usually want to see a turn key-solution that solves their problem.
3. The code that has been added to the code base makes the maintenance of the plugin harder. Moreover, it's also harder to parse and comprehend what is going on due to the increased complexity. I understand the convenience factor for users but it comes at a cost. I think the problem could have been solved in a different fashion without even adding code to the plugin (as mentioned above).

There are some additional points I'd like to raise about `CompositeExecInstruction`:

* We have to make sure that incremental build functionality works properly. Right now there's no test verifying the behavior.
* `CompositeExecInstruction` should rather live with the plugin implementation than the `Dockerfile` implementation. It's very specific to the plugin implementation and the use case it's trying to solve.
* I am not sure if I would have explicitly exposed the methods `defaultCommand` and `entryPoint` for `CompositeExecInstruction`. When I read the class name I'd expect to be able to add any type of instructions. Maybe the naming is just not concrete enough.
* Currently, the plugin cannot be used from a build script that uses Kotlin DSL. We have to turn the Groovy-specific APIs (Closures) into something that can be used with Groovy and Kotlin (Action). The exposed `exec` method uses a Closure in its signature meaning that there's an additional method that needs consideration and work.

I might be overly picky but it simply adds cognitive load when working on this code base. When I looked at it from a user's _and_ a plugin developer's perspective it wasn't clear to me what problem it solves.

**Comment by bmuschko** [comment]
One thing I'd like to mention: Would it maybe just be good enough if you'd provide your custom entrypoint when starting the container with [docker run --entrypoint run.sh](https://docs.docker.com/engine/reference/run/#entrypoint-default-command-to-execute-at-runtime)? The only thing you'd need to add to the image is the `run.sh` file.

**Comment by mwhipple** [comment]
Changing the docker run invocation seems tricky to plug into different schedulers when compared to having it be part of the image. If the concerns are mostly around current state I could likely help out with that (though vision for the project outweighs that). The current closure approach should maprobably to a Kotlin function with a receiver since it's not relying on runtime metaprogramming (the apply method written was intended to be equivalent to Kotlin).

**Comment by bmuschko** [comment]
> Changing the docker run invocation seems tricky to plug into different schedulers when compared to having it be part of the image.

Thanks for the info!

> If the concerns are mostly around current state I could likely help out with that (though vision for the project outweighs that).

Cool, I will probably work on clarifying the documentation soon.
